### ISSUE TITLE
Attribute: Changelog

### ISSUE THREAD

**Issue by michenriksen** [ReconJSON/ReconJSON#6]
## Description 
An attribute that contains a sort of changelog array that tools would add entries to when they create the initial file or when they decorate the document further with new data:

```
{
  "changelog": [
    ["2018-06-25T 2:57:00+0000", "amass v2.3.0"],
    ["2018-06-25T 3:23:00+0000", "aquatone v1.0.0"]
  ],
   ...
}
```

The above example would indicate that amass created the initial file that aquatone later consumed and decorated with further data.

## Reasoning
The changelog makes it easy to see which tools have already been used for a particular assessment in order to not duplicate work. Having a timestamp for when tools were used when is also useful for penetration testers who need to document their actions more thoroughly.

**Comment by Rhynorater** [comment]
Great thoughts, Michael. We should certainly have some sort of timestamp on the objects to tell when and by whom they were modified. 

What are your thoughts on how long the changelog should be? Should there be multiple entries from one tool?

**Comment by Rhynorater** [comment]
@michenriksen Also, what are your thoughts on whether the changelog should be on a file level, or on an object level?

**Comment by caffix** [comment]
I can see users being interested in the change log at the file level and for each object. Users also like to know the data source of each object. In this case, they would probably like to know the tool and data source that last updated the object, and the time stamp for when that happened.

**Comment by codingo** [comment]
Having done some work with graphing @caffix, would you want the ability to re-render a single object (knowing it had been updated) or would you tend to be cautious and just re-render the entire file?

I'm inclined to think a file update would be acceptable but also wonder if there's a piece to the puzzle I'm not seeing having not written anything graphing related before.

**Comment by Ice3man543** [comment]
@Rhynorater @michenriksen From what i understand by reading the issue, people will be interested in changelog at both file level and object level. I think this can be sorted out easily. 
For example, we could have a file level object describing the full changelog containing the actions performed where each action has a UUID associated to it. No two UUID should be the same. 

Here, at the end of the file there is a generic changelog structure having three fields associated with it. In code, it can be like this - 
```golang
type changelog struct {
    ID              string `json:"id",
    Timestamp string `json:"timestamp",
    Info             string `json:"info"
}
```
```javascript
  ...
  },
  {
      "changelog": [
           ["15e40d04-791e-11e8-adc0-fa7ae01bbebc", "2018-06-25T 2:57:00+0000", "subfinder v1.0.0"],
           ["15e40f70-791e-11e8-adc0-fa7ae01bbebc", "2018-06-25T 3:23:00+0000", "aquatone v1.0.0"]
      ],
  }
]
```

Then for each host object, we could add that ID specify what ran it. Sort-of like this in pseudo-approach
```javascript
{
    "uuid":["15e40d04-791e-11e8-adc0-fa7ae01bbebc", "15e40f70-791e-11e8-adc0-fa7ae01bbebc"],
    "type":"Host",
    "fqdn":"example.acme.com",
}
```

Now, there was the question of how to identify if multiple tools had worked on the same object. Which is why i decided to keep UUID field an array. This is my proposal for the solution to this changelog. 
I have few ideas about what is missing. But i don't think it is necessary to associate each object with a UUID. So, in my opinion the current approach is ok-ish.

If anyone has other ideas, please comment and let me know :+1:

**Comment by 0xdevalias** [comment]
While I do like the idea of knowing what data/updates came from which tool, etc; I wonder if this is starting to get into the 'bit too complex' case for the most basic thing. If the spec becomes too complex, I expect tool developers will be less and less inclined to use it. The way I see it, this should make things simpler (less thinking, etc) than not using it, so that people choose it for their tools.

So maybe this use case could be supported in some form of optional extension or similar, but I definitely don't think it belongs in the 'base'/core.

**Comment by Rhynorater** [comment]
To be honest, I agree with @0xdevalias. I think what this standard really needs to be is NOT the best, most thorough, "has every detail" standard. What it needs to be is something simple, not scary, and universal that developers can expand upon if the need be. ReconJSON should be extensible for things like a change log (perhaps we build this into suggestions for a database-like branch off), but should be fully usable by the community without one. Does that make sense?

For example, say @michenriksen wants a change log. He can implement that in Aquatone. However, when the ReconJSON JSON structure is imported into Subfinder which doesn't support the changelog, Subfinder should be able to import that data and use it seamlessly. Then, if the user so decides, they can wrap Subfinder to create the changelog. The idea of ReconJSON is interconnectivity and we won't get that by adding extraneous features, I don't think. Thoughts @michenriksen @Ice3man543 @codingo @0xdevalias?

**Comment by Rhynorater** [comment]
Although, I say, if we do anything, I feel like something like this would be helpful (proposed by Jobert) in issue #10. 
```
Given that the data that'll be stored using the specification is volatile, I'd think it'd be useful to have a generic set of attributes that give some insight into when the information was discovered, like last_observed_at which would contain an ISO 8601 date/time notation.
```

**Comment by Rhynorater** [comment]
@deadbits I agree - We'll make this change. If you like, you can create a pull request making the changes (please change to Camel Case). Otherwise, I will make the changes soon.

**Comment by 0xdevalias** [comment]
> @deadbits If certain apps don't support a given field, it doesn't really matter since they won't even know of their existence when loading / parsing a JSON file.

Definitely agree with this, and overall I think SOME form of date/time metadata would be useful. My point was more around considering what features we choose to add to the spec, as each time we make it more complicated, it makes it less and less likely that the implementing dev/application is likely to choose the spec, versus just doing their own thing (the "too hard, won't bother" problem)

**Comment by Rhynorater** [comment]
This has been addressed in #21. @deadbits is in charge of this feature.
